Mathlib/Analysis/Convex/Mul.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Analysis.Convex.Function
+import Mathlib.Algebra.Order.Monovary
+import Mathlib.Tactic.FieldSimp
+
+/-!
+# Product of convex functions
+
+This file proves that the product of convex functions is convex, provided they monovary.
+
+As corollaries, we also prove that `x ↦ x ^ n` is convex
+* `Even.convexOn_pow`: for even `n : ℕ`.
+* `convexOn_pow`: over $[0, +∞)$ for `n : ℕ`.
+* `convexOn_zpow`: over $[0, +∞)$ For `n : ℤ`.
+-/
+
+open Set
+
+variable {𝕜 E F : Type*}
+
+section LinearOrderedCommRing
+variable [LinearOrderedCommRing 𝕜] [LinearOrderedCommRing E]
+  [LinearOrderedAddCommGroup F] [Module 𝕜 E] [Module 𝕜 F] [Module E F] [IsScalarTower 𝕜 E F]
+  [SMulCommClass 𝕜 E F] [OrderedSMul 𝕜 E] [OrderedSMul 𝕜 F] [OrderedSMul E F] {s : Set 𝕜}
+  {f : 𝕜 → E} {g : 𝕜 → F}
+
+lemma ConvexOn.smul' (hf : ConvexOn 𝕜 s f) (hg : ConvexOn 𝕜 s g) (hf₀ : ∀ ⦃x⦄, x ∈ s → 0 ≤ f x)
+    (hg₀ : ∀ ⦃x⦄, x ∈ s → 0 ≤ g x) (hfg : MonovaryOn f g s) : ConvexOn 𝕜 s (f • g) := by
+  refine ⟨hf.1, fun x hx y hy a b ha hb hab ↦ ?_⟩
+  dsimp
+  refine (smul_le_smul (hf.2 hx hy ha hb hab) (hg.2 hx hy ha hb hab) (hg₀ $ hf.1 hx hy ha hb hab) $
+    add_nonneg (smul_nonneg ha $ hf₀ hx) $ smul_nonneg hb $ hf₀ hy).trans ?_
+  calc
+      _ = (a * a) • (f x • g x) + (b * b) • (f y • g y) + (a * b) • (f x • g y + f y • g x) := ?_
+    _ ≤ (a * a) • (f x • g x) + (b * b) • (f y • g y) + (a * b) • (f x • g x + f y • g y) :=
+        add_le_add_left (smul_le_smul_of_nonneg (hfg.smul_add_smul_le_smul_add_smul hx hy) $
```

Comment:
This looks like the kind of thing `gcongr` could help with, maybe leaving the subgoal solved by `hfg.smul_add_smul_le_smul_add_smul hx hy`.